keps/sig-cli/4764-kubectl-watchlist/README.md

```diff
@@ -0,0 +1,743 @@
+
+At this stage, we will adopt a similar approach to `client-go`, using an environment variable to control whether
+to enable this experimental feature, and ensuring that it can revert to the previous behavior if it fails.
+
+<<[UNRESOLVED Will **watchlist** replace **list** completely]>>
+**watchlist** was first released in version 1.27 and is still in the alpha stage, with continuous improvements 
+over the subsequent releases. For a feature in such an early stage, it is too soon to say whether we will fully
+rely on it without other fallback mechanisms.
+
+In my opinion, due to the performance improvements, we should use **watchlist** by default instead of **list** in the
+long term. For the evolution strategy, we can follow the approach of `client-go` and address it in a later release.
+
+<<[/UNRESOLVED]>>
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+Given that there is a large collection in the control plane, a user wants to fetch that collection using `kubectl get`.
+
+As a user, I shouldn't need to worry about the underlying details. I should simply update the version of `kubectl`
+to gain the performance improvements for free.
+
+#### Story 2
+
+As a cluster administrator, after upgrading the cluster, I should observe a significant reduction in the requests
+penetrating into etcd on the monitoring dashboard. Ideally, the majority of the requests should be cached by the 
+apiserver, rather than penetrating into etcd.
+
+### Risks and Mitigations
+
+For users, there are no visible changes to the user interface. However, since different APIs are called, 
+some users with restricted permissions may encounter authorization issues. Specifically, users might need
+to be granted additional **watch** permissions for the corresponding resources to ensure this mechanism 
+works properly. Fortunately, we have a fallback mechanism in this place. Even if users do not have the 
+additional permission, the old behavior should still work properly. In the worst case, some extra API 
+calls might cause a slight delay.
+
+Regarding security, users who already have **list** permissions inherently have access to all objects of this type.
+Therefore, I don't believe this change will introduce any additional security issues.
+
+## Design Details
+
+Since support for watch lists was merged in version 1.27, the API server allows clients to include the 
+`sendInitialEvents` query parameter when performing **watch** operations. This parameter instructs the 
+kube-apiserver to first send the existing members of the collection as `ADDED` events before pushing subsequent 
+objects changed events. A special `BOOKMARK` event with annotation `k8s.io/initial-events-end=true` will be sent 
+at the end to indicate the completion of the initial object push.
+
+Therefore, where kubectl originally performed a **list** operation using `List`, we can implement a new
```

Comment:
Instead of managing the appropriate parameters in kubectl, you should rely on internal client-go implementation that can decide itself whether watchlist can be usef for a given request.
And you achieve it effectively by setting client-go feature gate.

Reply:
I agree with  that we should rely on internal `client-go` implementation to decide.  However, due to current implementation issue, kubectl cannot directly use this feature-gate, which is what this KEP attempts to solve.

as we can see https://github.com/kubernetes/kubernetes/blob/fe3772890f650f9bcf020b43dc5a51fab0fa17f4/staging/src/k8s.io/cli-runtime/pkg/resource/helper.go#L108, kubectl use `Do` and `Get` method to list resource, not `List`. so the auto switch mechanism in `client-go` seems will not work here.



I described the introduction of a new environment variable earlier because I thought kubectl needed an independent switch to control it. If this is not necessary, I completely agree to reuse the environment variable switch in `client-go` to achieve this purpose. I will update KEP soon later.

Reply:
I just updated the KEP to clarify this part. PTAL